src/main/java/com/redhat/devtools/lsp4ij/features/formatting/LSPEditorImprovementsTypedHandler.java

```diff
@@ -0,0 +1,132 @@
+/*******************************************************************************
+ * Copyright (c) 2025 Red Hat, Inc.
+ * Distributed under license by Red Hat, Inc. All rights reserved.
+ * This program is made available under the terms of the
+ * Eclipse Public License v2.0 which accompanies this distribution,
+ * and is available at http://www.eclipse.org/legal/epl-v20.html
+ *
+ * Contributors:
+ * Red Hat, Inc. - initial API and implementation
+ ******************************************************************************/
+
+package com.redhat.devtools.lsp4ij.features.formatting;
+
+import com.intellij.codeInsight.editorActions.TypedHandlerDelegate;
+import com.intellij.openapi.editor.CaretModel;
+import com.intellij.openapi.editor.Document;
+import com.intellij.openapi.editor.Editor;
+import com.intellij.openapi.fileTypes.FileType;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.util.text.StringUtil;
+import com.intellij.psi.PsiFile;
+import com.intellij.util.containers.ContainerUtil;
+import com.redhat.devtools.lsp4ij.LSPIJEditorUtils;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Improves editor-typed handling of nested quote characters and statement terminators.
+ */
+public class LSPEditorImprovementsTypedHandler extends TypedHandlerDelegate {
+
+    @NotNull
+    @Override
+    public Result beforeCharTyped(char charTyped,
+                                  @NotNull Project project,
+                                  @NotNull Editor editor,
+                                  @NotNull PsiFile file,
+                                  @NotNull FileType fileType) {
+        if (LSPIJEditorUtils.isSupportedAbstractFileTypeOrTextMateFile(file)) {
+            if (LSPIJEditorUtils.isQuoteCharacter(file, charTyped) && handleNestedQuote(file, editor, charTyped)) {
+                return Result.STOP;
+            }
+            // TODO: Get statement terminator character(s) from client config?
+            else if ((';' == charTyped) && handleStatementTerminator(file, editor)) {
```

Comment:
...and statement terminator characters. Right now it's hard-coded for `;` as the only known statement terminator character. If there are known LSP-supported languages that have other statement terminator characters, we could instead make this configurable, but I'm not aware of any, and I didn't want to add that level of complexity without a concrete reason to do so.